charms/metallb-controller/src/charm.py

```diff
@@ -21,93 +23,93 @@
 logger = logging.getLogger(__name__)
 
 
-class MetallbControllerCharm(CharmBase):
+class MetalLBControllerCharm(CharmBase):
     """MetalLB Controller Charm."""
 
     _stored = StoredState()
 
     def __init__(self, *args):
         """Charm initialization for events observation."""
         super().__init__(*args)
-        if not self.model.unit.is_leader():
-            self.model.unit.status = WaitingStatus("Waiting for leadership")
+        if not self.unit.is_leader():
+            self.unit.status = WaitingStatus("Waiting for leadership")
             return
         self.image = OCIImageResource(self, 'metallb-controller-image')
-        self.framework.observe(self.on.start, self.on_start)
+        self.framework.observe(self.on.install, self._on_start)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.leader_elected, self._on_start)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.remove, self.on_remove)
+        self.framework.observe(self.on.remove, self._on_remove)
         # -- initialize states --
+        self._stored.set_default(k8s_objects_created=False)
         self._stored.set_default(started=False)
-        self._stored.set_default(configured=False)
+        self._stored.set_default(config_hash=self._config_hash())
         # -- base values --
         self._stored.set_default(namespace=os.environ["JUJU_MODEL_NAME"])
 
-    def _on_config_changed(self, event):
-        if not self._stored.started:
+    def _config_hash(self):
+        data = json.dumps({
```

Comment:
Why is this needed instead of grabbing directly the model config ?

Reply:
Ah, I imagine to compare all the config items instead of comparing them one by one to the stored object

Reply:
Yeah, this is more future-proofing for when we want to handle changes to the `protocol` config as well.  I also just generally consider it cleaner to not store a full copy of the data just to tell if it changed.